core/capabilities/remote/dispatcher.go

```diff
@@ -0,0 +1,146 @@
+	p2ptypes "github.com/smartcontractkit/chainlink/v2/core/services/p2p/types"
+)
+
+// dispatcher en/decodes messages and routes traffic between peers and capabilities
+type dispatcher struct {
+	peerWrapper p2ptypes.PeerWrapper
+	peer        p2ptypes.Peer
+	peerID      p2ptypes.PeerID
+	signer      p2ptypes.Signer
+	registry    commontypes.CapabilitiesRegistry
+	receivers   map[key]remotetypes.Receiver
+	mu          sync.RWMutex
+	stopCh      services.StopChan
+	wg          sync.WaitGroup
+	lggr        logger.Logger
+}
+
+type key struct {
+	capId string
+	donId string
+}
+
+var _ services.Service = &dispatcher{}
+
+const supportedVersion = 1
+
+func NewDispatcher(peerWrapper p2ptypes.PeerWrapper, signer p2ptypes.Signer, registry commontypes.CapabilitiesRegistry, lggr logger.Logger) *dispatcher {
+	return &dispatcher{
+		peerWrapper: peerWrapper,
+		signer:      signer,
+		registry:    registry,
+		receivers:   make(map[key]remotetypes.Receiver),
+		stopCh:      make(services.StopChan),
+		lggr:        lggr.Named("Dispatcher"),
+	}
+}
+
+func (d *dispatcher) Start(ctx context.Context) error {
+	d.peer = d.peerWrapper.GetPeer()
+	d.peerID = d.peer.ID()
+	if d.peer == nil {
+		return fmt.Errorf("peer is not initialized")
+	}
+	d.wg.Add(1)
+	go d.receive()
+	d.lggr.Info("dispatcher started")
+	return nil
+}
+
+func (d *dispatcher) SetReceiver(capabilityId string, donId string, receiver remotetypes.Receiver) error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	k := key{capabilityId, donId}
+	_, ok := d.receivers[k]
+	if ok {
+		return fmt.Errorf("receiver already exists for capability %s and don %s", capabilityId, donId)
+	}
+	d.receivers[k] = receiver
+	return nil
+}
+
+func (d *dispatcher) RemoveReceiver(capabilityId string, donId string) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	delete(d.receivers, key{capabilityId, donId})
+}
+
+func (d *dispatcher) Send(peerID p2ptypes.PeerID, msgBody *remotetypes.MessageBody) error {
+	msgBody.Version = supportedVersion
+	msgBody.Sender = d.peerID[:]
+	msgBody.Receiver = peerID[:]
+	rawBody, err := proto.Marshal(msgBody)
+	if err != nil {
+		return err
+	}
+	signature, err := d.signer.Sign(rawBody)
+	if err != nil {
+		return err
+	}
+	msg := &remotetypes.Message{Signature: signature, Body: rawBody}
+	rawMsg, err := proto.Marshal(msg)
+	if err != nil {
+		return err
+	}
+	return d.peer.Send(peerID, rawMsg)
+}
+
+func (d *dispatcher) receive() {
+	defer d.wg.Done()
+	recvCh := d.peer.Receive()
+	for {
+		select {
+		case <-d.stopCh:
+			d.lggr.Info("stopped - exiting receive")
+			return
+		case msg := <-recvCh:
+			body, err := ValidateMessage(msg, d.peerID)
+			if err != nil {
+				d.lggr.Errorw("received invalid message", "error", err)
+				continue
+			}
+			k := key{body.CapabilityId, body.DonId}
+			d.mu.RLock()
+			receiver, ok := d.receivers[k]
+			d.mu.RUnlock()
+			if !ok {
+				d.lggr.Debugw("received message for unregistered capability", "capabilityId", k.capId, "donId", k.donId)
```

Comment:
Hmmmm... so if a node receives a message for a capability that it doesn't support, does that mean we just don't respond to the client?

Reply:
Yeah, I agree that it might be better to respond with an error - let me add an enum to Message proto.
A counter-argument is not wasting resources on potentially buggy or malicious nodes spamming with invalid requests. But that will be addressed by rate-limiting anyway.